app/forms/csv_upload_form.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require 'csv'
+
+class CSVUploadForm
+  include ActiveModel::Validations
+
+  attr_reader :file
+
+  validates_with CSVValidator
+
+  def initialize(user)
+    @user = user
+  end
+
+  def save(params)
+    @file = params[:csv_upload_form][:file]
+
+    return false unless valid?
+
+    begin
+      Keyword.transaction do
+        # rubocop:disable Rails/SkipsModelValidations
+        user.keywords.insert_all parsed_keywords
```

Comment:
I think we need the bang method here
```suggestion
        user.keywords.insert_all! parsed_keywords
```

Without `!` method, it will not raise any exception and rails database translation will not work. What do you think?

Reply:
good insight :) 
I've check at the [documentation](https://apidock.com/rails/ActiveRecord/Persistence/ClassMethods/insert_all%21), I believe we do not need the bang here:
The bang only adds this:
> Raises ActiveRecord::RecordNotUnique if any rows violate a unique index on the table. In that case, no rows are inserted.
To skip duplicate rows, see ActiveRecord::Persistence#insert_all. To replace them, see ActiveRecord::Persistence#upsert_all.

In this situation, there is no risk of duplicates: Each ID is incrementally generated by the DB. If a user upload a CSV file with 20 same keywords, I believe he wants to compare how regular google ads works, in which case we will consider these 20 keywords as different.

Regarding exception catching for data validation, as the tests show, the DB is still throwing errors and no single keyword is being created if a file contains any invalid keyword. This was a strong focus point during my tests ;-)

So I think it is good without the bang. What do you think @junan ? :)

Reply:
My actual concern was with rails `transaction` block where(inside the block) we are not raising any exception manually or using `!`(which raises an expectation when it can't create a record rather than just returning false) when creating records.  And the purpose of `transaction` is to create all records or don't create any record if one record is failed to create and for that we usually need to raise an exception when one record is failed to create and it is easier to use bang `!` method to raise an exception rather than manually raise an exception.

I checked the code why it is working even you have not used `!` or raising manually and exception for this case.  I found that `insert_all! / insert_all ` skips validations(model validation), then it raises database null constraints exception if you insert invalid data because we set `null: false` constraints on `name` field in the database. That's why it was working as it raises exceptions in the database level.

Regarding the `transaction` example

This bellow code will not work as it does not raising an exception and end up creating first user.
```ruby
Keyword.transaction do
  [FFaker::Internet.email, ""].each do |email|
    User.create(last_name: FFaker::Name.last_name,
                first_name: FFaker::Name.first_name,
                email: email,
                password: 'password123',
                password_confirmation: 'password123')
  end
end
```

And if you raise exception using `!`(or manually) then it will work as it will not create any record
```ruby
Keyword.transaction do
  [FFaker::Internet.email, ""].each do |email|
    User.create!(last_name: FFaker::Name.last_name,
                first_name: FFaker::Name.first_name,
                email: email,
                password: 'password123',
                password_confirmation: 'password123')
  end
end
```
As `insert_all / insert_all! ` skips the validation and database raising the exception, I don't see also big benefit of using `!` in this case. So I am fine with both versions (bang or non bang). Its up to you which one you use chose

Reply:
yes exactly, we share the same understanding :) 
Thanks for all these details, it confirmed what I read. 
So as we said earlier: if both bang/no-bang are fine, let's prefer the no-bang ;-)